firedrake/functionspaceimpl.py

```diff
@@ -403,15 +409,19 @@ def reconstruct(self, mesh=None, name=None, **kwargs):
 
         if mesh is None:
             mesh = V_parent.mesh()
+        if element is None:
+            element = V_parent.ufl_element()
 
-        element = V_parent.ufl_element()
         cell = mesh.topology.ufl_cell()
         if len(kwargs) > 0 or element.cell != cell:
             element = element.reconstruct(cell=cell, **kwargs)
 
         V = type(self).make_function_space(mesh, element, name=name)
         for i in reversed(indices):
             V = V.sub(i)
+
+        if boundary_set:
+            V = RestrictedFunctionSpace(V, boundary_set=boundary_set, name=V.name)
```

Comment:
For separate discussion: should `RestrictedFunctionSpace` be hidden from the user?